docs/arrays.md

```diff
@@ -0,0 +1,291 @@
+        [2.48, 9.26],
+        [9.41, 7.26],
+        [7.45, 8.85],
+        [6.58, 5.22],
+        [6.05, 5.77],
+        [8.57, 4.16],
+        [3.22, 4.98],
+        [6.62, 6.62],
+        [9.36, 7.44],
+        [8.34, 3.43],
+        [4.47, 7.61],
+        [4.34, 5.05],
+        [5.0, 5.05],
+        [5.0, 5.0],
+        [2.07, 7.8],
+        [9.45, 9.6],
+        [3.1, 3.26],
+        [4.37, 5.72],
+    ],
+    dtype=pl.Array(pl.Float64, 2),
+)
+df = pl.DataFrame(points)
+
+print(df)
+```
+
+```
+shape: (22, 1)
+┌───────────────┐
+│ points        │
+│ ---           │
+│ array[f64, 2] │
+╞═══════════════╡
+│ [6.63, 8.35]  │
+│ [7.19, 4.85]  │
+│ [2.1, 4.21]   │
+│ [3.4, 6.13]   │
+│ [2.48, 9.26]  │
+│ …             │
+│ [5.0, 5.0]    │
+│ [2.07, 7.8]   │
+│ [9.45, 9.6]   │
+│ [3.1, 3.26]   │
+│ [4.37, 5.72]  │
+└───────────────┘
+```
+
+What if we wanted to make a plugin that takes a Series like `points` above, and likewise, returned a Series of 2d arrays?
+Turns out we _can_ do it! But it's a little bit tricky.
+
+Let's create a plugin that calculates the midpoint between a reference point and each point in a Series like the one above.
+This should illustrate both how to unpack an array inside our rust code and also return a Series of the same type.
+
+As usual, we start by registering our plugin:
+
+```python
+def midpoint_2d(expr: IntoExpr, ref_point: tuple[float, float]) -> pl.Expr:
+    return register_plugin_function(
+        args=[expr],
+        plugin_path=Path(__file__).parent,
+        function_name="midpoint_2d",
+        is_elementwise=True,
+        kwargs={"ref_point": ref_point},
+    )
+```
+
+As you can see, we included an additional kwarg: `ref_point`, which we annotated with the type `tuple: [float, float]`.
+In our rust code, we won't receive it as a tuple, though, it'll also be an array.
+This isn't crucial for this example, so just accept it for now.
+As you saw in the __arguments__ chapter, we take kwargs by defining a struct for them:
+
+```rust
+#[derive(Deserialize)]
+struct MidPoint2DKwargs {
+    ref_point: [f64; 2],
+}
+```
+
+And we can finally move to the actual plugin code:
+
+```rust
+#[polars_expr(output_type_func=point_2d_output)]
+fn midpoint_2d(inputs: &[Series], kwargs: MidPoint2DKwargs) -> PolarsResult<Series> {
+
+    let ca: &ArrayChunked = inputs[0].array()?;
+    let ref_point = kwargs.ref_point;
+
+    let out = unsafe {
+        ca.apply_amortized_same_type(|row| {
+            let s = row.as_ref();
+            let ca = s.f64().unwrap();
+            let out_inner: Float64Chunked;
+            match (ca.get(0), ca.get(1)) {
+                (Some(_), Some(_)) => {
+                    out_inner = ca
+                        .into_no_null_iter()
+                        .enumerate()
+                        .map(|(idx, val)| {
+                            let midpoint = (val + ref_point[idx]) / 2.0f64;
+                            Some(midpoint)
+                        }).collect_trusted();
```

Comment:
it looks like we're repeating work twice here? we're doing `ca.get(0)` and `ca.get(1)`, and then again iterating over `ca`

can we write this so that `ca` is only iterated over once?

Reply:
Yes, this is indeed the draft version, I managed to improve it to the version below, just didn't push it yet as I'd have to re-write the text which breaks it down line-by-line:
```rust
let out = unsafe {
    ca.apply_amortized_same_type(|row| {
        let s = row.as_ref();
        let ca = s.f64().unwrap();
        let out_inner: Float64Chunked = ca
            .into_no_null_iter()
            .enumerate()
            .map(|(idx, val)| {
                let midpoint = (val + ref_point[idx]) / 2.0f64;
                Some(midpoint)
            }).collect_trusted();
        out_inner.into_series()
    })};
```
I was waiting to talk to you before refactoring, as even with this version I might've missed a few things.

Reply:
nice!

a few comments
- `let ca = s.f64().unwrap();` might panic - can we avoid this? `try_apply_amortized_same_type`?
- `into_no_null_iter` assumes there's no nulls, which I think might not be the same here? why not use `apply_values` and preserve null values in the input?

Reply:
an alternative to `try_apply_amortized_same_type` would be to check the dtype before starting to iterate over `ca`, see https://github.com/MarcoGorelli/polars-plugins-tutorial/pull/47 for an example